python/node_similarity.py

```diff
@@ -0,0 +1,285 @@
+
+    :return: This procedure returns a mgp.Record with 3 fields:
+    the first node, the second node and the cosine similarity between them
+    :rtype: mgp.Record
+    """
+
+    return _calculate_similarity(node1, node2, _calculate_cosine, mode)
+
+
+def _calculate_jaccard(node1: mgp.Vertex, node2: mgp.Vertex) -> float:
+    """This method calculates the Jaccard similarity between two nodes.
+
+    :param node1: The first node
+    :type node1: mgp.Vertex
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return similarity: The Jaccard similarity of the first and the second node
+    :rtype: float
+    """
+
+    jaccard_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+    else:
+        jaccard_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    intersection_len = len(neighbors1 & neighbors2)
+
+    denominator = len(neighbors1) + len(neighbors2) - intersection_len
+
+    if denominator != 0:
+        jaccard_similarity = intersection_len / denominator
+
+    return jaccard_similarity
+
+
+def _calculate_overlap(node1: mgp.Vertex, node2: mgp.Vertex) -> float:
+    """This method calculates the overlap similarity between two nodes.
+
+    :param node1: The first node
+    :type node1: mgp.Vertex
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return: The overlap similarity of the first and the second node
+    :rtype: float
+
+    """
+
+    overlap_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+    else:
+        overlap_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    denominator = min(len(neighbors1), len(neighbors2))
+
+    if denominator != 0:
+        overlap_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return overlap_similarity
+
+
+def _calculate_cosine(node1: mgp.Vertex, node2: mgp.Vertex) -> float:
+    """This method calculates the cosine similarity between two nodes.
+
+    :param node1: The first node
+    :type node1: mgp.Vertex
+    :param node2: The second node
+    :type node2: mgp.Vertex
+
+    :return: The cosine similarity of the first and the second node
+    :rtype: float
+    """
+
+    cosine_similarity: float
+
+    neighbors1: set
+    neighbors2: set
+
+    if node1 == node2:
+        return 1.0
+    else:
+        cosine_similarity = 0.0
+
+    neighbors1 = _get_neighbors(node1)
+    neighbors2 = _get_neighbors(node2)
+
+    denominator = sqrt(len(neighbors1) * len(neighbors2))
+
+    if denominator != 0:
+        cosine_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return cosine_similarity
+
+
+def _calculate_similarity(
+    node1: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    node2: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    method: Callable,
+    mode: str,
+) -> List[Tuple[mgp.Vertex, mgp.Vertex, float]]:
```

Comment:
I am not an expert but I think 4 values as parameters are a little bit too much, think if you can somehow reduce it.

Reply:
Maybe mode can be something like default value.

Reply:
I don't see a problem with 4 arguments.